Python/gc_free_threading.c

```diff
@@ -298,6 +299,34 @@ gc_visit_heaps(PyInterpreterState *interp, mi_block_visit_fun *visitor,
     return err;
 }
 
+static void
+gc_visit_thread_stacks(struct _stoptheworld_state *stw)
+{
+    HEAD_LOCK(&_PyRuntime);
```

Comment:
Let's refactor this as something like the following to break up the deeply nested loops.

Also, I think it's worth adding a helper like `PyStackRef_IsDeferred`.

```c
static void
gc_visit_frame(_PyInterpreterFrame *frame)
{
    PyCodeObject *co = _PyFrame_GetCode(frame);
    for (int i = 0; i < co->co_nlocalsplus + co->co_stacksize; i++) {
        _PyStackRef curr_o = frame->localsplus[i];
        // Note: we MUST check that it has deferred bit set before checking the rest.
        // Otherwise we might read into invalid memory due to non-deferred references
        // being dead already.
        if (PyStackRef_IsDeferred(curr_o)) {
            PyObject *obj = PyStackRef_Get(curr_o);
            if (!_Py_IsImmortal(obj)) {
                gc_add_refs(obj, 1);
            }
        }
    }
}

static void
gc_visit_thread_stacks(PyInterpreterState *interp)
{
    HEAD_LOCK(&_PyRuntime);
    for (PyThreadState *p = interp->threads.head; p != NULL; p = p->next) {
        _PyInterpreterFrame *curr_frame = p->current_frame;
        while (curr_frame != NULL) {
            // f_executable could be Py_None for the entry frame.
            if (curr_frame->f_executable != Py_None) {
                gc_visit_frame(curr_frame);
            }
            curr_frame = curr_frame->previous;
        }
    }
    HEAD_UNLOCK(&_PyRuntime);
}
```